README.md
# accesslogger


[![Documentation](https://godoc.org/github.com/mashiike/accesslogger?status.svg)](https://godoc.org/github.com/mashiike/accesslogger)
![Latest GitHub tag](https://img.shields.io/github/tag/mashiike/accesslogger.svg)
![Github Actions test](https://github.com/mashiike/accesslogger/workflows/Test/badge.svg?branch=main)
[![Go Report Card](https://goreportcard.com/badge/mashiike/accesslogger)](https://goreportcard.com/report/mashiike/accesslogger)
[![License](https://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/mashiike/accesslogger/blob/master/LICENSE)

accessloger for golang http handler

## Requirements
  * Go 1.18 or higher. support the 3 latest versions of Go.

## Usage 

sample go code
```go
package main

import (
	"encoding/json"
	"log"
	"net/http"
	"os"

	"github.com/mashiike/accesslogger"
)

func main() {
	err := http.ListenAndServe("localhost:8080", 
		accesslogger.Wrap(
			http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			w.WriteHeader(http.StatusOK)
				e := json.NewEncoder(w)
				e.SetEscapeHTML(true)
				e.SetIndent("", "  ")
				e.Encode(map[string]string{"status":"ok"})
			}),
			accesslogger.CombinedLogger(os.Stderr),
			//accesslogger.CombinedDLogger(os.Stderr),
			//accesslogger.JSONLogger(os.Stderr),
        ),
	)
	if err != nil {
		log.Fatalln(err)
	}
}
```

output log:
```
192.0.2.1:1234 - - [26/Dec/2022:15:04:05 +0900] "GET / HTTP/1.1" - 4 "-" "go test client"
192.0.2.1:1234 - hoge [26/Dec/2022:15:04:06 +0900] "GET /hoge HTTP/1.1" - 4 "https://example.com" "go test client"
222.222.333.333 - hoge [26/Dec/2022:15:04:06 +0900] "GET /hoge HTTP/1.1" - 4 "https://example.com" "go test client"
```


CHANGELOG.md
# Changelog

## [v0.2.0](https://github.com/mashiike/accesslogger/compare/v0.1.0...v0.2.0) - 2023-03-07
- add headers for custom log format by @mashiike in https://github.com/mashiike/accesslogger/pull/2

## [v0.1.0](https://github.com/mashiike/accesslogger/commits/v0.1.0) - 2022-12-21


middleware_test.go
package accesslogger_test

import (
	"bytes"
	"fmt"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"

	"github.com/mashiike/accesslogger"
	"github.com/sebdah/goldie/v2"
)

func TestWrap(t *testing.T) {
	var past time.Duration
	base, _ := time.Parse(time.RFC3339, "2022-12-26T15:04:05+09:00")
	accesslogger.Clock = func() time.Time {
		return base.Add(past)
	}

	var combinedLogs bytes.Buffer
	var combinedDLogs bytes.Buffer
	var jsonLogs bytes.Buffer
	h := accesslogger.Wrap(
		http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
			time.Sleep(10 * time.Millisecond)
			past += 10 * time.Millisecond
			fmt.Fprintf(w, "hoge")
		}),
		accesslogger.CombinedLogger(&combinedLogs),
		accesslogger.CombinedDLogger(&combinedDLogs),
		accesslogger.JSONLogger(&jsonLogs),
	)
	g := goldie.New(t)

	w := httptest.NewRecorder()
	r := httptest.NewRequest(http.MethodGet, "/", nil)
	r.Header.Set("User-Agent", "go test client")
	h.ServeHTTP(w, r)
	r.Header.Set("Referer", "https://example.com")
	r.URL.Path = "/hoge"
	r.SetBasicAuth("hoge", "fuga")
	past = 1 * time.Second
	h.ServeHTTP(w, r)
	r.Header.Set("CloudFront-Viewer-Address", "222.222.333.333")
	h.ServeHTTP(w, r)

	g.Assert(t, "combined", combinedLogs.Bytes())
	g.Assert(t, "combinedD", combinedDLogs.Bytes())
	g.Assert(t, "json", jsonLogs.Bytes())
}


logger.go
package accesslogger

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"
)

type AccessLog struct {
	RequestHeader  http.Header `json:"-"`
	ResponseHeader http.Header `json:"-"`
	RemoteAddr     string      `json:"remote_addr,omitempty"`
	AccessedAt     time.Time   `json:"accessed_at,omitempty"`
	UserAgent      string      `json:"user_agent,omitempty"`
	Referer        string      `json:"referer,omitempty"`
	BasicAuthUser  string      `json:"basic_auth_user,omitempty"`
	Request        string      `json:"request,omitempty"`
	StatusCode     int         `json:"status_code,omitempty"`
	BodyByteSent   int         `json:"body_byte_sent,omitempty"`
	FirstSentAt    time.Time   `json:"first_sent_at,omitempty"`
	LastSentAt     time.Time   `json:"last_sent_at,omitempty"`
	FirstSentTime  int64       `json:"first_sent_time,omitempty"`
	ResponseTime   int64       `json:"response_time,omitempty"`
}

func NewAccessLog(r *http.Request) *AccessLog {
	user, _, _ := r.BasicAuth()
	l := &AccessLog{
		RequestHeader: r.Header,
		RemoteAddr:    coalesce(r.Header.Get("CloudFront-Viewer-Address"), r.RemoteAddr),
		AccessedAt:    Clock(),
		UserAgent:     r.UserAgent(),
		Referer:       r.Referer(),
		BasicAuthUser: user,
		Request:       fmt.Sprintf("%s %s %s", r.Method, r.URL.Path, r.Proto),
	}
	return l
}

func (l *AccessLog) WriteResponseInfo(w *ResponseWriter) *AccessLog {
	l.StatusCode = w.StatusCode
	l.BodyByteSent = w.BodyByteSent
	l.FirstSentAt = w.FirstWriteTime
	l.LastSentAt = w.LastWriteTime
	l.FirstSentTime = l.FirstSentAt.Sub(l.AccessedAt).Microseconds()
	l.ResponseTime = l.LastSentAt.Sub(l.AccessedAt).Microseconds()
	l.ResponseHeader = w.Header()
	return l
}

type Logger interface {
	WriteAccessLog(*AccessLog)
}

type FormatLogger struct {
	io.Writer
	LogFormat
}

func (logger FormatLogger) WriteAccessLog(l *AccessLog) {
	fmt.Fprintln(logger.Writer, logger.LogFormat(l))
}

type LogFormat func(*AccessLog) string

func CombinedLogFormat(l *AccessLog) string {
	return fmt.Sprintf(
		`%s %s %s [%s] "%s" %s %s "%s" "%s"`,
		coalesce(l.RemoteAddr, "-"),
		"-",
		coalesce(l.BasicAuthUser, "-"),
		l.AccessedAt.Format("02/Jan/2006:15:04:05 -0700"),
		coalesce(l.Request, "-"),
		coalesce(emptyif(fmt.Sprintf("%d", l.StatusCode), "0"), "-"),
		coalesce(fmt.Sprintf("%d", l.BodyByteSent), "0"),
		coalesce(l.Referer, "-"),
		coalesce(l.UserAgent, "-"),
	)
}

func CombinedLogger(w io.Writer) FormatLogger {
	return FormatLogger{
		Writer:    w,
		LogFormat: CombinedLogFormat,
	}
}

func CombinedDLogFormat(l *AccessLog) string {
	return fmt.Sprintf(
		`%s %s`,
		CombinedLogFormat(l),
		coalesce(fmt.Sprintf("%d", l.ResponseTime), "0"),
	)
}

func CombinedDLogger(w io.Writer) FormatLogger {
	return FormatLogger{
		Writer:    w,
		LogFormat: CombinedDLogFormat,
	}
}

func JSONLogFormat(l *AccessLog) string {
	bs, err := json.Marshal(l)
	if err != nil {
		return `{}`
	}
	return string(bs)
}

func JSONLogger(w io.Writer) FormatLogger {
	return FormatLogger{
		Writer:    w,
		LogFormat: JSONLogFormat,
	}
}
